hawkbit-device-simulator/src/main/java/org/eclipse/hawkbit/simulator/amqp/DmfReceiverService.java

```diff
@@ -182,37 +185,89 @@ private void handleEventMessage(final Message message, final String thingId) {
-        final String tenant = (String) headers.get(MessageHeaderKey.TENANT);
+    private void handleMultiActionRequest(final Message message, final String thingId) {
+
+        final DmfMultiActionRequest multiActionRequest = convertMessage(message, DmfMultiActionRequest.class);
+        final String tenant = getTenant(message);
+
+        final DmfMultiActionRequest.DmfMultiActionElement actionElement =
+                multiActionRequest.getElements().get(0);
+
+        processAction(thingId, tenant, actionElement);
+    }
+
+    private void processAction(final String thingId, final String tenant,
+            final DmfMultiActionRequest.DmfMultiActionElement actionElement) {
+        final EventTopic eventTopic = actionElement.getTopic();
+        final DmfActionRequest action = actionElement.getAction();
+        final long actionId = action.getActionId();
+
+        if(openActions.stream().anyMatch(id -> id == actionId)) {
+            return;
+        }
+
+        openActions.add(actionId);
 
+        switch (eventTopic) {
+            case DOWNLOAD :
+            case DOWNLOAD_AND_INSTALL :
+                if (action instanceof DmfDownloadAndUpdateRequest) {
+                    processUpdate(thingId, eventTopic, tenant, (DmfDownloadAndUpdateRequest) action);
+                }
+                break;
+            case CANCEL_DOWNLOAD :
+                processCancelDownloadAction(thingId, tenant, action.getActionId());
+                break;
+            default :
+                LOGGER.info("No valid event property in MULTI_ACTION.");
+                break;
+        }
+    }
+
+    private void handleAttributeUpdateRequest(final Message message, final String thingId) {
+        final String tenant = getTenant(message);
         spSenderService.updateAttributesOfThing(tenant, thingId);
     }
 
-    private void handleCancelDownloadAction(final Message message, final String thingId) {
+    private static String getTenant(final Message message) {
         final MessageProperties messageProperties = message.getMessageProperties();
         final Map<String, Object> headers = messageProperties.getHeaders();
-        final String tenant = (String) headers.get(MessageHeaderKey.TENANT);
+        return (String) headers.get(MessageHeaderKey.TENANT);
+    }
+
+    private void handleCancelDownloadAction(final Message message, final String thingId) {
+        final String tenant = getTenant(message);
         final Long actionId = convertMessage(message, Long.class);
 
+        processCancelDownloadAction(thingId, tenant, actionId);
+    }
+
+    private void processCancelDownloadAction(final String thingId, final String tenant, final Long actionId) {
         final SimulatedUpdate update = new SimulatedUpdate(tenant, thingId, actionId);
-        spSenderService.finishUpdateProcess(update, Arrays.asList("Simulation canceled"));
+        spSenderService.finishUpdateProcess(update, Collections.singletonList("Simulation canceled"));
```

Comment:
do we not need to remove the action from the `openActions` collection here as well?